Typeck break expr even if break is illegal

gurry opened the pull request:
Fixes #117821

We were returning immediately when encountering an illegal break. However, this caused problems later when the expr that the break was returning was evaluated during writeback. So now we don't return and instead simply set tainted by error. This lets typeck of break expr to occur even though we've encountered an illegal break.

compiler-errors commented on compiler/rustc_hir_typeck/src/expr.rs:
```diff
@@ -626,18 +626,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 e_ty = self.check_expr_with_hint(e, coerce_to);
+
+                // If break with value was found illegal above,
+                // return error right here or we risk passing typeck
```
It doesn't matter if we pass typeck. We've already delayed a bug. What happens if you remove this?

compiler-errors replied:
Alternatively, you should taint the infcx in that `unwrap_or_else` above.

gurry replied:
> It doesn't matter if we pass typeck. We've already delayed a bug. What happens if you remove this?

If we pass typeck here it results in an ICE during MIR interpretation as reported in #114529. I had created PR #117382 to fix #114529. However, that resulted in regression #117821. The present PR intends to fix that regression.

gurry replied:
> Alternatively, you should taint the infcx in that `unwrap_or_else` above.

You're right, tainting infcx in `unwrap_or_else` is much cleaner. I've done that now.

gurry commented:
@compiler-errors please review

compiler-errors commented:
@gurry: It has been 3 days, 2 of which have been the weekend. Please be patient, I will review it when I have time :)

compiler-errors commented on compiler/rustc_hir_typeck/src/expr.rs:
```diff
@@ -626,15 +626,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                // Set expectation to error in that case and set tainted
+                // by error (#114529)
+                let coerce_to = opt_coerce_to.unwrap_or_else(|| {
+                    let err = Ty::new_misc_error(tcx);
+                    if let ty::Error(e) = err.kind() {
+                        self.set_tainted_by_errors(*e);
                     }
```
Please manually delay an error instead of using an if-let.

```suggestion
                    let guar = tcx.sess.delay_span_bug(...);
                    self.set_tainted_by_errors(guar);
                    Ty::new_error(tcx, guar)
```

gurry commented:
> @gurry: It has been 3 days, 2 of which have been the weekend. Please be patient, I will review it when I have time :)
> 
> I don't think it is necessary to ping a reviewer until it's been _at least_ a week without review.

Apologies. I keep forgetting that you are just one person reviewing a large volume of PRs. Will be mindful of that next time.

gurry committed:
Typeck break expr even if break is illegal

We were earlier returning immediately when encountering an illegal break. However, this caused problems later
when the expr that the break was returning was evaluated during writeback. So now we don't return and instead
simply set tainted by error. This lets typeck of break expr to occur even though we've encountered an illegal break.

gurry commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

compiler-errors commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 4657917f6e6c49893e4fbcbe71469d4489b743d0 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 4657917f6e6c49893e4fbcbe71469d4489b743d0 -->
<!-- homu: {"type":"Approved","sha":"4657917f6e6c49893e4fbcbe71469d4489b743d0","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
